atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java

```diff
@@ -287,6 +287,7 @@ private CassandraService clientPoolWithParams(
                         .build())
                 .localHostWeighting(weighting)
                 .consecutiveAbsencesBeforePoolRemoval(1)
+                .keyspace("ks")
```

Comment:
Interesting. I'd love to hear why this becomes necessary (I have some suspicions, but I'd like to see if they match your reasoning!)

Reply:
Essentially frontloading the exception to construction time, except in the case where `CassandraClientFactory` is created but `create` is never called, or `getClientInternal` is called but you don't have a keyspace set (in which it's a break, since you now need to pass in a keyspace). This is because `create` previously required a keyspace (or would throw)

 `getClientInternal` is called in `CCPI#sanityCheckRingConsistency`, but that method always requires a keyspace, and `CassandraVerifier`'s keyspace methods, where it also needs a keyspace.

I don't however know if `CassandraClientFactory` is created but never used outside of AtlasDB - a sourcegraph search showed no use. However, the tests never called create, and so they would now be throwing an exception where they would not be before.

This could be mitigated using a similar setup to the config itself (optional keyspace, getKeyspaceOrThrow) - OTOH I noticed CassandraAtlasDbFactory mandates a keyspace before giving you back the merged config (or would throw)

Reply:
My main point here is: evidently, we don't use the keyspace in these tests (else it'd throw); is it now the case that we use some derived config which _does_ need the keyspace (but still don't need the keyspace itself)? And thus, as you noted, is this ever going to be a problem anywhere? I think the main places to check other than here are atlasdb-proxy and internal atlas backup service, but it does sound like you've done some searching anyway.